capsules/src/net/sixlowpan.rs

```diff
@@ -602,166 +736,146 @@ impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
+    /// Sets the [SixlowpanClient](trait.SixlowpanClient.html) that will receive
+    /// transmission completion and new packet reception callbacks.
+    pub fn set_client(&'a self, client: &'a SixlowpanClient) {
+        self.client.set(Some(client));
     }
 
-    /// FragState::transmit_packet
-    /// --------------------------
-    /// This function is called to send a fully-formed IPv6 packet. Arguments
-    /// to this function are used to determine various aspects of the MAC
-    /// layer frame and keep track of the transmission state.
+    /// Transmits the supplied IPv6 packet.
+    ///
+    /// Transmitted IPv6 packets will be optionall compressed, fragmented and
+    ///
+    /// Only one transmission is allowed at a time. Calling this method while
+    /// before a previous tranismission has completed will return an error.
+    ///
+    /// # Arguments
+    ///
+    /// * `src_mac_addr` - Why is the argument specified?
+    ///
+    /// * `dst_mac_addr` - Why is the argument specified?
+    ///
+    /// * `ip6_packet` - A buffer containing the IPv6 packet to transmit. This
+    /// buffer will be passed back in the
+    /// [send_done](trait.SixlowpanClient.html#send_done) callback.
+    ///
+    /// * `ip6_packet_len` - The length of the packet, in case the `ip6_packet`
+    /// buffer is larger than the actual packet. Must be at most
+    /// `ip6_packet.len()`
+    ///
+    /// * `security` - An optional tuple specifying the security level
+    /// (encryption, MIC, or both) and key identifier to use. If elided, no
+    /// security is used.
+    ///
+    /// This function exposes the primary packet transmission fuctionality. The
+    /// source and destination mac address arguments specify the link-layer
+    /// addresses of the packet, while the `security`, `fragment`, and `compress`
+    /// options specify the security level (if any), whether to fragment the packet
+    /// if it is too large, and whether to compress the packet respectively. These
+    /// different options are exposed to provide the caller with more control over
+    /// how this layer modifies packets; no compression and no fragmentation can
+    /// be set, and the packet will be sent to the radio as a raw IPv6 packet.
+    ///
+    /// The `ip6_packet` argument contains a pointer to a buffer containing a valid
+    /// IPv6 packet, while the `ip6_packet_len` argument specifies the number of
+    /// bytes to send. Note that `ip6_packet.len() > ip6_packet_len`, but we check
+    /// the invariant that `ip6_packet_len <= ip6_packet.len()`.
     pub fn transmit_packet(&self,
                            src_mac_addr: MacAddress,
                            dst_mac_addr: MacAddress,
                            ip6_packet: &'static mut [u8],
                            ip6_packet_len: usize,
                            security: Option<(SecurityLevel, KeyId)>,
-                           tx_state: &'a TxState<'a>,
                            fragment: bool,
```

Comment:
Are `security`, `fragment` and `compress` something I might want to modify on a per-packet basis? Or should they, instead, be part of the `Sixlowpan` state (and someone will modify them either in the constructor or with setter methods)? Again, how is the sender suppose to decide what options to use?

The baseline signature is `transmit_packet(ip6_packet: &'static mut [u8]) -> Result<...>` and any additional arguments need to be justified (e.g. `ip6_packet_len` is needed because of a limitation we currently have with managing static buffers).

Reply:
So security is definitely changed on a per-packet basis, and some packets cannot be fragmented (and so `fragment` was exposed as an easy way to check/fail if the packet wouldn't fit).

`compress` was exposed largely for testing reasons I think, and that can probably be removed.

Reply:
> So security is definitely changed on a per-packet basis

Can you be more specific? Is it just that the initial MLE messages in Thread are unencrypted, but after connection establishment everything _is_, followed by periodic key rotations? If so, this all sounds like it should be part of the `Sixlowpan` state and changed "on the side" with specific setter methods.

I'm hearing that at least both `fragment` and `compress` should be removed. If there are cases where a packet can't be fragmented and we're better off failing, it should probably be a separate method. In general, a single method with all the options exposed and one or more helper methods with some or all left out is a reasonable design.

So how is something like the `UDP` driver meant to decide which security level to use, or whether to fragment?

Reply:
So @bbbert is more familiar with the security configuration, but some packets need link-layer encryption, and some packets do not. This is handled by the Thread specification, so I think that UDP should just forward that decision up the stack, and expose it in its `send` method. I think there is also additional per-packet state that needs to be encoded in the security object, but I'm not familiar enough with how Thread deals with 802.15.4 security to say for certain.